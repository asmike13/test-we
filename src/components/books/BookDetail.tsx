import BookInfos from "./BookInfos"
import NavBar from "../common/NavBar"
import { IBook } from "src/types/book"
import CharacterList from "../characters/containers/CharacterListContainer"
import Loader from "../common/Loader"

interface IBookDetailProps {
	status: "idle" | "fulfilled" | "pending" | "rejected"
	currentBook?: IBook
	title: string
}

export const BookDetail = ({ status, currentBook, title }: IBookDetailProps) => (
	<>
		{status === 'pending' && <Loader />}
		{status === 'rejected' && <div className="section-title">Book not found</div>}
		{status === 'fulfilled' && currentBook &&
			<>
				<NavBar title={title} />
				<span className="section-title">Infos :</span>
				<BookInfos book={currentBook} />
				<CharacterList />
			</>
		}
	</>
)

export default BookDetail;
