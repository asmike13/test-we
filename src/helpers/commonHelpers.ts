import { IBook } from "../types/book"
import { ICharacter } from "../types/character"

export const getBookId = (url: string) => url.split('books/')[1]

export const getCharacterId = (url: string) => url.split('characters/')[1]
export const getCharacterIds = (urls: string[]) => urls
	.map((url) => getCharacterId(url))
	// remove undefined ids
	.filter((url) => url)

export const getCharacterName = (character?: ICharacter) => character?.name || character?.aliases[0] || ''

export const getBookFromBookList = (bookList: IBook[], id: string) => bookList.find((book) => getBookId(book.url) === id)
export const getCharacterFromCharacterList = (characterList: ICharacter[], id?: string) => characterList.find((character) => getCharacterId(character.url) === id)
