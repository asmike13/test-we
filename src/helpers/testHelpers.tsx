/* eslint-disable prefer-const */

import { Provider } from "react-redux";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import { store } from "src/store";


export const getRender = (Component: () => JSX.Element, initRoute: string[] = ['/'], path?: string) => {
	return (
		<Provider store={store}>
			<MemoryRouter initialEntries={initRoute}>
				<Routes>
					<Route path="/" element={<Component />} />
					{path && <Route path={path} element={<Component />} />}
					{!path && [...initRoute].map((r, index) => (
						<Route key={index} path={r} element={<Component />} />
					))}
				</Routes>
			</MemoryRouter>
		</Provider>
	)
}

export const waitForData = (store: any, value: any) => new Promise((res) => {
	const loop = (s: any, v: any) => {
		if (JSON.stringify(store) === JSON.stringify(value)) {
			res(true)
		} else {
			setTimeout(() => {
				loop(store, value)
			}, 300);
		}
	}

	loop(store, value)
})

export const waitForElement = (
	container: HTMLElement,
	element: string,
	removed?: boolean,
	valueToEqual?: string,
	length?: number,
) => new Promise((resolve, reject) => {
	let loopTimeout: any;
	let timeout: any;
	let elementFound: any;

	const getTextContent = (el: any) => el?.textContent || el?.value;
	let isFoundAndEqual: any;

	const loop = () => {
		elementFound = valueToEqual
			? container.querySelectorAll(element)
			: container.querySelector(element);

		const isFound = elementFound && !removed && !valueToEqual;
		const isRemoved = !elementFound && removed;
		const isFoundAndLengthEqual = isFound && length && elementFound?.length === length;

		isFoundAndEqual = valueToEqual
			&& Object.values(elementFound).find((el: any) => getTextContent(el) === valueToEqual);

		if (isRemoved || isFound || isFoundAndEqual || isFoundAndLengthEqual) {
			clearTimeout(timeout);
			resolve(isFoundAndEqual || elementFound);
		} else {
			loopTimeout = setTimeout(() => {
				loop();
			}, 100);
		}
	};

	loop();

	timeout = setTimeout(() => {
		clearTimeout(loopTimeout);
		reject(Error(`${element} ${valueToEqual ? `expect ${valueToEqual} get ${getTextContent(Object.values(elementFound)[0])}` : 'not found'}`));
	}, 3000);
});

export const waitForElementToBeRemoved = (
	container: HTMLElement,
	element: string,
) => waitForElement(
	container,
	element,
	true,
);

export const waitForElementToEqual = (
	container: HTMLElement,
	element: string,
	valueToEqual: string,
) => waitForElement(
	container,
	element,
	undefined,
	valueToEqual,
);

export const waitForElementLengthToEqual = (
	container: HTMLElement,
	element: string,
	length: number,
) => waitForElement(
	container,
	element,
	undefined,
	undefined,
	length,
);
