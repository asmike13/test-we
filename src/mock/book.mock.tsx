import { IBook } from "src/types/book";

export const bookMock = {
	authors: ['string'],
	characters: ['https://anapioficeandfire.com/api/characters/1'],
	country: 'string',
	isbn: 'string',
	mediaType: 'string',
	name: 'string',
	numberOfPages: 1,
	povCharacters: ['https://anapioficeandfire.com/api/characters/1'],
	publisher: 'string',
	released: '1996-08-01T00:00:00',
	url: 'https://anapioficeandfire.com/api/books/1',
} as IBook

export const booksMock = [
	bookMock,
	{
		...bookMock,
		url: 'https://anapioficeandfire.com/api/books/2',
	},
]