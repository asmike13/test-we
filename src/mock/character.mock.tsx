import { ICharacter } from "src/types/character";

export const characterMock = {
	id: '1',
	url: 'https://anapioficeandfire.com/api/characters/1',
	name: 'string',
	gender: 'string',
	culture: 'string',
	born: 'string',
	died: 'string',
	titles: ['string'],
	aliases: ['string'],
	father: 'string',
	mother: 'string',
	spouse: 'string',
	allegiances: ['string'],
	books: ['https://anapioficeandfire.com/api/books/1'],
	povBooks: ['https://anapioficeandfire.com/api/books/1'],
	tvSeries: ['string'],
	playedBy: ['string'],
} as ICharacter

export const charactersMock = [
	characterMock,
	{
		...characterMock,
		url: 'https://anapioficeandfire.com/api/characters/2',
	}
]