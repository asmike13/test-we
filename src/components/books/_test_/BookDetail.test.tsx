import renderer from 'react-test-renderer';
import { bookMock } from 'src/mock/book.mock';
import BookDetail from '../BookDetail';
import { getRender } from "src/helpers/testHelpers"

it('should render as snapshot with status idle', async () => {
	const snapshot = renderer.create(getRender(() => <BookDetail status='idle' title="title" />))

	expect(snapshot).toMatchSnapshot()
})

it('should render as snapshot with status rejected', async () => {
	const snapshot = renderer.create(getRender(() => <BookDetail status='rejected' title="title" />))

	expect(snapshot).toMatchSnapshot()
})

it('should render as snapshot with status pending', async () => {
	const snapshot = renderer.create(getRender(() => <BookDetail status='pending' title="title" />))

	expect(snapshot).toMatchSnapshot()
})

it('should render as snapshot with status fulfilled', async () => {
	const snapshot = renderer.create(getRender(() => <BookDetail status='fulfilled' title="title" currentBook={bookMock} />))

	expect(snapshot).toMatchSnapshot()
})
