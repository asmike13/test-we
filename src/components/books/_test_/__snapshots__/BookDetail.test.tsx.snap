// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render as snapshot with status fulfilled 1`] = `
Array [
  <header
    className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
  >
    <div
      className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="HomeIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
          />
        </svg>
      </a>
      <a
        className="arrow-back"
        href="/"
        onClick={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ArrowBackIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
          />
        </svg>
      </a>
      <h1
        className="MuiTypography-root MuiTypography-h1 navbar-title css-sqoje6-MuiTypography-root"
      >
        title
      </h1>
    </div>
  </header>,
  <span
    className="section-title"
  >
    Infos :
  </span>,
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 book-infos css-obvova-MuiPaper-root"
  >
    <ul
      className="MuiList-root MuiList-padding css-xpcooq-MuiList-root"
    >
      <li
        className="MuiListItem-root MuiListItem-gutters css-1xtwiqb-MuiListItem-root"
        disabled={false}
      >
        <div
          className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="NoteAltIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3h-4.18C14.4 1.84 13.3 1 12 1s-2.4.84-2.82 2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-7-.25c.41 0 .75.34.75.75s-.34.75-.75.75-.75-.34-.75-.75.34-.75.75-.75zM9.1 17H7v-2.14l5.96-5.96 2.12 2.12L9.1 17zm7.75-7.73-1.06 1.06-2.12-2.12 1.06-1.06c.2-.2.51-.2.71 0l1.41 1.41c.2.2.2.51 0 .71z"
            />
          </svg>
        </div>
        By string
      </li>
      <li
        className="MuiListItem-root MuiListItem-gutters css-1xtwiqb-MuiListItem-root"
        disabled={false}
      >
        <div
          className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AutoStoriesIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m19 1-5 5v11l5-4.5V1zM1 6v14.65c0 .25.25.5.5.5.1 0 .15-.05.25-.05C3.1 20.45 5.05 20 6.5 20c1.95 0 4.05.4 5.5 1.5V6c-1.45-1.1-3.55-1.5-5.5-1.5S2.45 4.9 1 6zm22 13.5V6c-.6-.45-1.25-.75-2-1v13.5c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5v2c1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.3 4.75 1.05.1.05.15.05.25.05.25 0 .5-.25.5-.5v-1.1z"
            />
          </svg>
        </div>
        1 pages
      </li>
      <li
        className="MuiListItem-root MuiListItem-gutters css-1xtwiqb-MuiListItem-root"
        disabled={false}
      >
        <div
          className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="LocalPrintshopIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 8H5c-1.66 0-3 1.34-3 3v6h4v4h12v-4h4v-6c0-1.66-1.34-3-3-3zm-3 11H8v-5h8v5zm3-7c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1zm-1-9H6v4h12V3z"
            />
          </svg>
        </div>
        Published by string
      </li>
      <li
        className="MuiListItem-root MuiListItem-gutters css-1xtwiqb-MuiListItem-root"
        disabled={false}
      >
        <div
          className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CalendarMonthIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 4h-1V2h-2v2H8V2H6v2H5c-1.11 0-1.99.9-1.99 2L3 20c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 16H5V10h14v10zM9 14H7v-2h2v2zm4 0h-2v-2h2v2zm4 0h-2v-2h2v2zm-8 4H7v-2h2v2zm4 0h-2v-2h2v2zm4 0h-2v-2h2v2z"
            />
          </svg>
        </div>
        Released on 01/08/1996
      </li>
    </ul>
  </div>,
  <div
    className="character-list"
  >
    <h2
      className="section-title"
    >
      Characters :
    </h2>
    <div
      className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 css-3su884-MuiPaper-root"
    >
      <ul
        className="MuiList-root MuiList-padding css-epc8ws-MuiList-root"
        id="character-scroll-list"
      />
    </div>
  </div>,
]
`;

exports[`should render as snapshot with status idle 1`] = `null`;

exports[`should render as snapshot with status pending 1`] = `
Array [
  <div
    className="overlay"
  />,
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation5 loader css-qh7kl5-MuiPaper-root"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="HourglassBottomIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="m18 22-.01-6L14 12l3.99-4.01L18 2H6v6l4 4-4 3.99V22h12zM8 7.5V4h8v3.5l-4 4-4-4z"
      />
    </svg>
  </div>,
]
`;

exports[`should render as snapshot with status rejected 1`] = `
<div
  className="section-title"
>
  Book not found
</div>
`;
